### PR TITLE
Adding support for simultaneous REST requests across multiple users

### PR DESCRIPTION
**Pull request by bhariharan**

### CODE REVIEW

**Comment by huminzhi** [comment]
## Danger has errored
[!] Invalid `DangerFile` file: `minimum_coverage_percentage` value must be a Float! Found Integer instead.
```
 #  from DangerFile:5
 #  -------------------------------------------
 #  
 >  xcov.report(
 #     scheme: 'UnitTests',
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: hybrid/SampleApps/NoteSync/NoteSync/SFContentSoqlSyncDownTarget.m [hunk 1]
```diff
@@ -194,7 +194,7 @@ - (id)initWithQueryLocator:(NSString*)queryLocator
     return self;
 }
 
-- (NSURLRequest *)prepareRequestForSend
+- (NSURLRequest *)prepareRequestForSend:(SFUserAccount *)user
 {
     NSString *sessionId = [SFAuthenticationManager sharedManager].coordinator.credentials.accessToken;
```
Comment (trooper2013) [inline hybrid/SampleApps/NoteSync/NoteSync/SFContentSoqlSyncDownTarget.m]: We should use user.credentials.accessToken instead.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 2]
```diff
@@ -40,16 +39,15 @@
 NSString* const kSFDefaultContentType = @"application/json";
 NSInteger const kSFRestErrorCode = 999;
 
-// singleton instance
-static SFRestAPI *_instance;
-static dispatch_once_t _sharedInstanceGuard;
 static BOOL kIsTestRun;
+static NSMutableDictionary *sfRestApiList = nil;
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: Couldn't we replace this with SFSDKSafeMutableDictionary instead? Then all the guarded   blocks can be reduced?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 3]
```diff
@@ -108,11 +106,45 @@ - (void)cancelAllRequests {
...
-    return _instance;
+    return [SFRestAPI sharedInstanceWithUser:[SFUserAccountManager sharedInstance].currentUser];
+}
+
++ (SFRestAPI *)sharedInstanceWithUser:(SFUserAccount *)user {
+    static dispatch_once_t pred;
+    dispatch_once(&pred, ^{
+        sfRestApiList = [[NSMutableDictionary alloc] init];
+    });
+    @synchronized ([SFRestAPI class]) {
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: See previous comment about  SFSDKSafeMutableDictionary.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 4]
```diff
@@ -108,11 +106,45 @@ - (void)cancelAllRequests {
...
+        if (!sfRestApi) {
+            sfRestApi = [[SFRestAPI alloc] initWithUser:user];
+            sfRestApiList[key] = sfRestApi;
+        }
+        return sfRestApi;
+    }
+}
+
++ (void)removeSharedInstanceWithUser:(SFUserAccount *)user {
+    @synchronized ([SFRestAPI class]) {
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: See previous comment about  SFSDKSafeMutableDictionary.

**Comment by bhariharan** [comment]
@trooper2013 Done [here](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/2180/commits/0acec595a5aecd0e3e9e0b4d52a5692c941c676d).

**Comment by radhap** [comment]
Can you please squash all the commits into 1, so it gets little easy for cherry-picking ? Thanks in advance

**Pull request merged**